StripeFinancialConnections/StripeFinancialConnections/Source/Native/Shared/CustomSheet/CustomSheetViewController.swift

```diff
@@ -0,0 +1,206 @@
+
+    func dismissSheet(completionHandler: (() -> Void)? = nil) {
+        sheetDidDismissClosure = completionHandler
+        dismissSheetClosure?()
+    }
+
+    fileprivate func sheetDidDismiss() {
+        sheetDidDismissClosure?()
+    }
+}
+
+extension UIViewController {
+
+    @available(iOSApplicationExtension, unavailable)
+    func presentAsSheet(communicationHelper: CustomSheetCommunicationHelper) {
+        PresentCustomSheet(
+            type: .contentViewController(self),
+            sheetCommunicationHelper: communicationHelper
+        )
+    }
+}
+
+private enum CustomSheetPresentationType {
+    case contentView(UIView)
+    case contentViewController(UIViewController)
+}
+
+@available(iOSApplicationExtension, unavailable)
+private func PresentCustomSheet(
+    type: CustomSheetPresentationType,
+    sheetCommunicationHelper: CustomSheetCommunicationHelper
+) {
+    guard let topMostViewController = UIViewController.topMostViewController() else {
+        assertionFailure("Expected to always find a top view controller.")
+        return
+    }
+
+    // First present an invisible view controller over the full screen.
+    // After, when we present a sheet over the invisible view controller,
+    // the invisible view controller will give an appearance as if
+    // the iOS standard sheet animation _does not_ visually modify
+    // the view hierarchy with a "hierarchy/depth" animation.
+    let invisibleViewController = InvisibleViewController()
+    invisibleViewController.modalPresentationStyle = .overFullScreen
+    topMostViewController.present(invisibleViewController, animated: false)
+
+    let didDissmiss: () -> Void = { [weak invisibleViewController] in
+        // when the sheet VC dismisses, so will the `invisibleViewController`
+        invisibleViewController?.dismiss(
+            animated: false,
+            completion: {
+                sheetCommunicationHelper.sheetDidDismiss()
+            }
+        )
+    }
+
+    let customSheetViewController: CustomSheetViewController
+    switch type {
+    case .contentViewController(let contentViewController):
+        customSheetViewController = CustomSheetViewController(
+           contentViewController: contentViewController,
+           didDismiss: didDissmiss
+       )
+    case .contentView(let contentView):
+        customSheetViewController = CustomSheetViewController(
+           contentView: contentView,
+           didDismiss: didDissmiss
+       )
+    }
+    sheetCommunicationHelper.dismissSheetClosure = { [weak customSheetViewController] in
+        customSheetViewController?.dismiss(animated: true)
+    }
+    invisibleViewController.present(customSheetViewController, animated: true)
+}
+
+// TODO(kgaidis): for iOS 13, we may need to improve sheet background
+@available(iOSApplicationExtension, unavailable)
+private final class CustomSheetViewController: UIViewController, UIGestureRecognizerDelegate {
```

Comment:
I may have to re-write this later because if the sheet is smaller than the size of the window, there is a little bit of a weird "shadow" that is noticeable when presenting or slowly dismissing the sheet. That being said, I should be able to swap it out relatively easily later.